2025/build-tooling/demo/final/src/App.tsx

```diff
@@ -1,12 +1,194 @@
+import Graphic from "@arcgis/core/Graphic";
+import WebStyleSymbol from "@arcgis/core/symbols/WebStyleSymbol";
+import Point from "@arcgis/core/geometry/Point";
+import Collection from "@arcgis/core/core/Collection";
+import { webMercatorToGeographic } from "@arcgis/core/geometry/support/webMercatorUtils";
+
+import type { ApiKeyManager } from "@esri/arcgis-rest-request";
+import { findPlacesNearPoint } from "@esri/arcgis-rest-places";
+
+type Nil<T> = T | null | undefined;
+
+interface Result<T> {
+  result?: T;
+  error?: Error;
+  loading?: true;
+}
+
+interface ServiceInfo {
+  authentication: ApiKeyManager;
+  endpoint?: string;
+}
+
+const schoolSymbol = new WebStyleSymbol({
+  name: "school",
+  styleName: "Esri2DPointSymbolsStyle",
+});
+
+export function App({ serviceInfo }: { serviceInfo: ServiceInfo }) {
+  const featuresElement = useRef<HTMLArcgisFeaturesElement>(null);
+  const [mapElement, setMapElement] = useState<Nil<HTMLArcgisMapElement>>();
+  const [selectedFeature, setSelectedFeature] = useState<Nil<Graphic>>();
+  const [schoolResults, setSchoolResults] =
+    useState<Result<Collection<Graphic>>>();
+
+  const featureActions = useMemo<HTMLArcgisFeaturesElement["actions"]>(
+    () =>
+      new Collection([
+        {
+          title: "Load nearby schools",
+          id: "load-schools",
+          icon: "education",
+        },
+      ]),
+    [],
+  );
+
+  const onMapClick = useCallback(
+    (event: HTMLArcgisMapElement["arcgisViewClick"]) => {
+      featuresElement.current?.open({
+        location: event.detail.mapPoint,
+        fetchFeatures: true,
+      });
+    },
+    [],
+  );
+
+  const onFeaturesChange = useCallback(
+    (event: HTMLArcgisFeaturesElement["arcgisPropertyChange"]) => {
+      if (event.detail.name === "selectedFeature") {
+        setSelectedFeature(event.target.selectedFeature);
+        setSchoolResults(undefined);
+      }
+    },
+    [],
+  );
+
+  const onFeatureActionClicked = useCallback(
+    async (event: HTMLArcgisFeaturesElement["arcgisTriggerAction"]) => {
+      const { action } = event.detail;
+      if (action.id !== "load-schools" || !selectedFeature?.geometry) return;
+
+      // Convert the selected feature's geometry to lat/long
+      const latLngGeometry = webMercatorToGeographic(selectedFeature.geometry);
+      const latLngCenter = (latLngGeometry as Polygon).centroid!;
+
+      // Find schools near the selected feature
+      let response;
+      try {
+        setSchoolResults({ loading: true });
+        response = await findPlacesNearPoint({
+          x: latLngCenter.x,
+          y: latLngCenter.y,
+          radius: 1600 * 5, // 5 miles in meters
+          categoryIds: ["4d4b7105d754a06372d81259"], // Schools
+          authentication: serviceInfo.authentication,
+          endpoint: serviceInfo.endpoint,
+        });
+      } catch (error) {
+        if (error instanceof Error) {
+          setSchoolResults({ error });
+        }
```

Comment:
```suggestion
        setSchoolResults({ error: String(error) })
```

Reply:
Currently the interface expects the actual error, this way it can render different strings dependent on the error content (although currently it just shows a generic error)